### PR TITLE
Updated the 8-bit SoftmaxOp exponent calculation with 4 16-bit tables to avoid extreme large/small slope errors

### PR DESCRIPTION
**Pull request by Jerry-Ge**
Softmax uses 2x16-bit table lookups to calculate the exp() value. With the stricter checking on the interpolation of 16-bit table, the table lookups are failing. Instead, I updated the softmax to use 4x16-bit tables to do the exp() calculation which can avoid failing test cases when the interpolation slope is too large/small.

Signed-off-by: Jerry Ge <jerry.ge@arm.com>

### CODE REVIEW

**Comment by Jerry-Ge** [comment]
@jpienaar Hi Jacques, here comes my second PR which fixes one of the SoftmaxOp issue stated above, could you please help review this one when available? Thanks!

**Comment by Jerry-Ge** [comment]
cc @eric-k256

**Review by jpienaar — COMMENTED** [review]
I'm surprised no test needed updating. Is there execution tests for these?

**Comment by Jerry-Ge** [comment]
Thanks for the reply! Great question, since this is not a new legalization for Softmax which means it's only necessary to pass the existing SoftmaxOp legalization tests. > I'm surprised no test needed updating. Is there execution tests for these?

**Comment by eric-k256** [comment]
I'd expect test_softmax_qi8 in tensorflow/compiler/mlir/tosa/tests/tfl-to-tosa-pipeline.mlir to fail with this change. The test code itself is fine, but the CHECK lines will need to be updated to match what the new softmax code is generating. The non int8 softmax should continue to work as is.

**Comment by Jerry-Ge** [comment]
> I'd expect test_softmax_qi8 in tensorflow/compiler/mlir/tosa/tests/tfl-to-tosa-pipeline.mlir to fail with this change. The test code itself is fine, but the CHECK lines will need to be updated to match what the new softmax code is generating. The non int8 softmax should continue to work as is.

Good catch! I will update that accordingly.

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc [hunk 1]
```diff
@@ -1438,70 +1439,97 @@ llvm::Optional<Value> convertSoftmaxOp(PatternRewriter& rewriter, Operation* op,
...
-      auto op15_lshift_op12_op14 = CreateOpAndInfer<tosa::LogicalLeftShiftOp>(
+      auto op18_lshift_op15_op17 = CreateOpAndInfer<tosa::LogicalLeftShiftOp>(
           rewriter, op->getLoc(), int32_rsum_type,
-          op12_reducesum_op11.getResult(), op14_sub_op13.getResult());
+          op15_reducesum_op14.getResult(), op17_sub_op16.getResult());
 
       // Step 5. Calculate one_over_one_plus_x() with Newton-Raphson division
       // with 3 iterations.
       // Need two magic constants 48/17 and -32/17 from Newton-Raphson algorithm
       // We need to operator in s2.29 since 48/17 is > 2.0
```
[the remark this answers is not in the document]
Comment (Jerry-Ge) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:1530]: Yup..Will update that as well :)

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc [hunk 2]
```diff
@@ -1438,70 +1439,97 @@ llvm::Optional<Value> convertSoftmaxOp(PatternRewriter& rewriter, Operation* op,
...
-          exp_table_const_upper);
+          exp_table_const_01);
 
-      auto op6_table_op4_lower = CreateOpAndInfer<tosa::TableOp>(
+      auto op6_table_op4_02 = CreateOpAndInfer<tosa::TableOp>(
           rewriter, op->getLoc(), int32_logits_type, op4_rescale_op3,
-          exp_table_const_lower);
+          exp_table_const_02);
+
+      auto op7_table_op4_03 = CreateOpAndInfer<tosa::TableOp>(
```
[the remark this answers is not in the document]
Comment (Jerry-Ge) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:1450]: Hmm..Let me think

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc [hunk 3]
```diff
@@ -1438,70 +1439,97 @@ llvm::Optional<Value> convertSoftmaxOp(PatternRewriter& rewriter, Operation* op,
...
 
-      auto op12_reducesum_op11 = CreateOpAndInfer<tosa::ReduceSumOp>(
+      auto op15_reducesum_op14 = CreateOpAndInfer<tosa::ReduceSumOp>(
           rewriter, op->getLoc(), int32_rsum_type,
-          op11_rshift_op10_12.getResult(),
+          op14_rshift_op13_12.getResult(),
           rewriter.getI64IntegerAttr(input_rank - 1));
 
       // Step 4. calculate reciprocal(sum(exp()))
       // CLZ returns headroom_plus_one
```
[the remark this answers is not in the document]
Comment (Jerry-Ge) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:1512]: Yup.. Will update that accordingly.

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jerry-Ge) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:1450]: How about `op7_table_op4_first8`,  `op7_table_op4_second8`, etc. something like those?

**Comment by Jerry-Ge** [comment]
@jpienaar Hi Jacques, I updated the code followed by your review comments yesterday. Could you please help take another look when possible. Tks!

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpienaar) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc]: The opN prefix means small changes end up with a lot of renaming, which I hope we can avoid in future. I don't know if we could have something like "first_order_lookup_table" or "newton_raphson_bits8t16" or something. I just don't get an intuition of what op7_table_op4_first8 is (the first8 is an improvement, table is useful too even if more what it is than what it represents).

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc [hunk 4]
```diff
@@ -1438,70 +1439,97 @@ llvm::Optional<Value> convertSoftmaxOp(PatternRewriter& rewriter, Operation* op,
...
-      auto op10_add_op7_op9 = CreateOpAndInfer<tosa::AddOp>(
+      // Add together all the 8-bit groups
+      // Add [1+2]
+      auto op11_add_op7_op8 = CreateOpAndInfer<tosa::AddOp>(
           rewriter, op->getLoc(), int32_logits_type, op7_lshift_op5.getResult(),
-          op9_add_op8_32768.getResult());
+          op8_lshift_op6.getResult());
+
+      // Add [1+2+3]
+      auto op12_add_op14_op9 = CreateOpAndInfer<tosa::AddOp>(
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:1491]: Isn't this an add of op11 and op9 ?

**Review by jpienaar — APPROVED** [review]
Seems good, I didn't verify the numerics as much, did you verify on reference implementation that this matches?

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jerry-Ge) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc]: I totally agree with that and it's quite tedious when updating those opN. I am currently using the naming like `op8_table_op4_bits_7_0` and `op7_table_op4_bits_15_8`, etc. For all the opN names changes here, I believe it could be another PR which is solely for refactoring this to avoid the confusion between the original intention of this PR? What do you think? : )

[the remark this answers is not in the document]
Comment (jpienaar) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc]: SGTM

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc (hunk 4 above)
Comment (Jerry-Ge) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:1491]: Good catch. I think it's a consist typo here. I will update this quickly.

Comment (Jerry-Ge) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:1491]: The results are good.

**Review by jpienaar — APPROVED** [review]

**Comment by Jerry-Ge** [comment]
It weird that it fails the mlir tests while I passed all the them locally..

**Comment by Jerry-Ge** [comment]
It's weird that it fails the MLIR tests in Py+CPP Test Suite while I passed all the them on my local machine..

**Comment by jpienaar** [comment]
> It's weird that it fails the MLIR tests in Py+CPP Test Suite while I passed all the them on my local machine..

Did you figure it out?

**Comment by Jerry-Ge** [comment]
> > It's weird that it fails the MLIR tests in Py+CPP Test Suite while I passed all the them on my local machine..
> 
> Did you figure it out?

I read the error outputs and it seems there VAR#s are off by 1 a little bit but I can pass all the them locally. I made the smaller update and already pushed that. Waiting for the updated test results here

**Comment by Jerry-Ge** [comment]
@jpienaar It fails the `ARM CI Extended / build (3.10) (pull_request) ` test where I saw a lot of FLAKY keywords there. Is there anyway to re-run those tests and could I have the permission to trigger the test runs here for efficiency? Tks

**Comment by jpienaar** [comment]
Triggering another run, I can try and submitting from inside if this is just flake (easier for me to rerun things there)

**Comment by jpienaar** [comment]
Mmm that reran many tests but not that one

**Comment by Jerry-Ge** [comment]
> Mmm that reran many tests but not that one

Yes..I feel it's a little bit unstable for those tests scripts here. I'm still figuring out why I can pass it locally but not here for the MLIR failures from Py+CPP here. 😂 (A little bit hard to debug in such case without being able to run..)

**Comment by Jerry-Ge** [comment]
Taking a note here: the large constant values has to be written like `0x5{{.*}}` in the mlir test files, otherwise it causes the pattern match failures.

**Review by jpienaar — APPROVED** [review]

**Comment by Jerry-Ge** [comment]
Alright, it seems the `Py+CPP` passed here. While `Windows Bazel GPU — Internal CI build failed
`  is failing everyone else which I believe it's not due to my fault?

**Comment by jpienaar** [comment]
Its going through another round of presubmits now, if all goes well should be done post.
